File: app/src/main/java/net/ginteam/carmen/presenter/category/CategoriesPresenter.java
package net.ginteam.carmen.presenter.category;

import net.ginteam.carmen.contract.category.CategoriesContract;
import net.ginteam.carmen.model.category.CategoryModel;
import net.ginteam.carmen.provider.ModelCallback;
import net.ginteam.carmen.provider.category.CategoriesProvider;

import java.util.List;

/**
 * Created by Eugene on 12/23/16.
 */

public class CategoriesPresenter implements CategoriesContract.Presenter {

    private CategoriesContract.View mView;

    @Override
    public void selectCategory(CategoryModel category) {

    }

    @Override
    public void fetchCategories() {
        mView.showLoading(true);

        CategoriesProvider
                .getInstance()
                .fetchCategories(new ModelCallback<List<CategoryModel>>() {
                    @Override
                    public void onSuccess(List<CategoryModel> resultModel) {
                        mView.showLoading(false);
                        mView.showCategories(resultModel);
                    }

                    @Override
                    public void onFailure(String message) {
                        mView.showError(message);
                    }
                });
    }

    @Override
    public void attachView(CategoriesContract.View view) {
        mView = view;
    }

    @Override
    public void detachView() {
        mView = null;
    }

}

File: app/src/main/kotlin/net/ginteam/carmen/kotlin/presenter/category/CategoriesPresenter.kt
package net.ginteam.carmen.kotlin.presenter.category

import net.ginteam.carmen.kotlin.api.response.ModelSubscriber
import net.ginteam.carmen.kotlin.contract.CategoriesContract
import net.ginteam.carmen.kotlin.manager.PreferencesManager
import net.ginteam.carmen.kotlin.manager.SharedPreferencesManager
import net.ginteam.carmen.kotlin.model.CategoryModel
import net.ginteam.carmen.kotlin.presenter.BasePresenter
import net.ginteam.carmen.kotlin.provider.CategoryDataSourceProvider
import net.ginteam.carmen.kotlin.provider.OnlineCategoryDataSourceProvider

/**
 * Created by eugene_shcherbinock on 2/16/17.
 */
class CategoriesPresenter : BasePresenter <CategoriesContract.View>(), CategoriesContract.Presenter {

    private val mCategoriesProvider: CategoryDataSourceProvider = OnlineCategoryDataSourceProvider()

    override fun fetchCategories() {
        mView?.showLoading(true)

        val preferences: PreferencesManager = SharedPreferencesManager
        mCategoriesProvider
                // it can not be null in this place
                .fetchCategories(preferences.userCityModel!!.id)
                .subscribe(object : ModelSubscriber <List <CategoryModel>>() {
                    override fun success(model: List<CategoryModel>) {
                        mView?.showLoading(false)
                        mView?.showCategories(model)
                    }

                    override fun error(message: String, isNetworkError: Boolean) {
                        mView?.showError(message, isNetworkError)
                    }
                })
    }

}